### DISCUSSION TITLE
EdgeDB slow when connected to a remote database

### DISCUSSION THREAD

**Discussion by clmnin** [geldata/gel#2457] [Q&A]
I tried setting up a LightSail Postgres Database and two LightSail instances - one for EdgeDB and one for my python app.

<p align="center">
<img src="https://user-images.githubusercontent.com/17064666/114934009-fd5c6900-9e56-11eb-82fd-a0181a7be686.png">
</p>

Commands used to load test EdgeDB 

1. Create a database in a remote pg cluster
    ```sh
    # connect to RDS database endpoint
    psql --host=<rds database endpoint>--port=5432 --username=<username> --password --dbname=postgres
    # create a database in rds
    create database test;
    ```
2. Connect to rds using edb
    ```sh
    # use edb
    edb server --postgres-dsn "postgres://<username>:<password>@<rds database endpoint>:5432/test" --no-testmode --port=5656 --bind-address 0.0.0.0
    ```
3. Run python app locally
    ```sh
    python app/server/server.py
    ```
4. Load test the app using Locust
    ```sh
    locust --host=http://localhost -u=400 -r=10 --headless
    ```

## Observations
1. When EdgeDB in docker or in a local pg cluser (edb server --postgres-dsn postgres://postgres:me@localhost/test) my load test could handle 200 requests per second (rps)
2. When pointing to a remote cluser the rps drops drastically to 2-5 requests per second.

## Questions
1. How should I configure `edb` so that I can use it to connect to a remote pg cluser?
2. How does @fantix do EdgeDB automated tests? Do you have multiple workers running multiple tests on the same edgedb + pg cluser?

**Comment by elprans** [comment]
Is this on beta.1 or nightly?

**Reply by clmnin** [reply]
1.0-beta.2+dev.5568.gc8c562284 
Build from source as per [this](https://www.edgedb.com/docs/internals/dev) doc

**Reply by elprans** [reply]
Please attach an EdgeDB server log from the benchmark run, if possible.

**Reply by clmnin** [reply]
It looks like I have the same issues as #2428 . Here is the log file
[edb.log](https://github.com/edgedb/edgedb/files/6323505/edb.log)

The log file was created using
```sh
edb server --postgres-dsn "postgres dsn" --no-testmode --port=5656 --bind-address 0.0.0.0 --log-level=debug --echo-runtime-info --log-to edb.log
```

DatabaseConnectionView object has no attribute 'name'

I'm using
1. Lightsail RDS instance of 1GB RAM, 1vCPU, 40GB SSD
2. Lightsail instance 1GB RAM, 1vCPU, 40GB SSD

**Reply by fantix** [reply]
@clmno I fixed #2428 in #2464, but I'm not so sure it is the root reason to the issue you met. Do you mind also sharing the code that serves the load test in `app/server/server.py`, or even better - a self-contained example please?

**Reply by clmnin** [reply]
@fantix @elprans I'll share a self-contained example soon, been a bit busy at work.

Sorry for the delay in responses. We are in totally different time zones.

**Comment by fantix** [comment]
> 1. How should I configure `edb` so that I can use it to connect to a remote pg cluser?

If my understanding of your question is correct, you're doing it the right way - by providing the `--postgres-dsn` parameter. FWIW you could provide some (or all) of the sensitive parameter through environment variable, like:

```console
$ PGHOST=<rds database endpoint> PGUSER=<username> PGPASSWORD=<password> edb server --postgres-dsn=postgres://
```

(Reference [here](https://www.postgresql.org/docs/current/libpq-envars.html) - this is what EdgeDB tries to be compatible to)

> 2. How does @fantix do EdgeDB automated tests? Do you have multiple workers running multiple tests on the same edgedb + pg cluser?

The test suite is running on a single machine (GitHub Action worker) with something like this:

```console
$ edb test --postgres-dsn=<Postgres DSN> -j2
```

This starts one EdgeDB testing serer (connecting to one remote RDS) and 2 worker processes to run tests in parallel. But some tests may start their own tempoaray EdgeDB servers connecting to the same remote RDS.
